File: C++ Solutions/DesignAStackWithIncrementOperation.cpp
#include <vector>
#include <iostream>

class CustomStack
{
private:
    std::vector<int> stack, inc;
    int maxSize;

public:
    CustomStack(int maxSize)
    {
        this->maxSize = maxSize;
    }

    void push(int x)
    {
        if (this->stack.size() == this->maxSize)
            return;
        this->stack.push_back(x);
        this->inc.push_back(0);
    }

    int pop()
    {
        int i = this->stack.size() - 1;

        if (i < 0)
            return -1;
        if (i > 0)
            this->inc[i - 1] += this->inc[i];

        int res = this->stack[i] + this->inc[i];

        this->stack.pop_back();
        this->inc.pop_back();

        return res;
    }

    void increment(int k, int val)
    {
        int i = std::min(k, (int)this->stack.size()) - 1;
        if (i >= 0)
            this->inc[i] += val;
    }
};

File: TypeScript Solutions/DesignAStackWithIncrementOperation.ts
class CustomStack {
    private stack: number[];
    private maxSize: number;

    public constructor(maxSize: number) {
        this.stack = [];
        this.maxSize = maxSize;
    }

    public push(x: number): void {
        this.stack.length < this.maxSize && this.stack.push(x);
    }

    public pop(): number {
        return this.stack.length ? this.stack.pop() : -1;
    }

    public increment(k: number, val: number): void {
        const it = k < this.stack.length ? k : this.stack.length;

        for(let i=0; i<it; i++) {
            this.stack[i] += val;
        }
    }
}
